### PR TITLE
Migrate re-ordering to stimulus controller w-sortable

### PR DESCRIPTION
**Pull request by salty-ivy**
Fixes #https://github.com/wagtail/wagtail/issues/10909

migrating the re-ordering logic from https://github.com/wagtail/wagtail/blob/main/wagtail/admin/templates/wagtailadmin/pages/index.html#L79-L208

into a stimulus controller `w-sortable` ( suggested name ) via incorporating [sortable.js](https://github.com/SortableJS/Sortable)

CC: @lb-   

https://github.com/wagtail/wagtail/assets/74553951/efc8a1bf-923f-4f3e-bbc0-159c44f8ebe9

_Please check the following:_

-   [ ] Do the tests still pass?[^1]
-   [ ] Does the code comply with the style guide?
    -   [ ] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by salty-ivy** [comment]
@lb-  as of now this works perfectly with `mouse only interactions` although there some hard-coded values like `draggable` class ( which is causing `table-tests` failure at backend )  which  I will fix later when few things finalizes.

let me know your thoughts over this approach and any ideas about how can we deal with `keyboard` only interaction with this approach.

**Comment by lb-** [comment]
Wow. This is great. I'll take a deep dive look next week.

**Review by lb- — COMMENTED** [review]
This is a great start, I could not get this working as is but I understand it's just a draft for now.

Here's a bunch of feedback, hopefully it helps you continue down this path. I think we may need a few rounds of back and forth (if you are up for it). Thank you though for wanting to get this together.

1. Let's use a consistent set of naming `w-sortable` for the controller identifier, `SortableController` for the controller and `SortableController.ts` for the file name. Shifting between Sort/Sortable in naming will cause confusion later.
2. Set up a new target `handle` and then use `data-w-sortable-target="handle"` for all handles. This would replace the `.handle` setting / class. Then, when setting up sortable, pass in the selector for this attribute `[data-${this.identifier}-target="handle"]`. This means that a. we are not using classes for JS selection and b. we can leverage the handleTargets if needed later.
3. In a similar way let's use `item` as the target for the draggable items. This way we would set `data-w-sortable-target="item"` and then pass in `[data-${this.identifier}-target="item"]` to the Sortable instance for `draggable`.
4. Remove `wagtail/contrib/modeladmin/static/wagtailmodeladmin/js/prepopulate.js` this has been added accidentally. It's just a build artifact from other changes that have been made recently that can just be deleted.
5. I think we should revisit the way we are posting to the server, we can lean in more on actual HTML here. I think we can better leverage one of the existing ActionController or SubmitController to handle this async request for each move (maybe). For now though, let's continue as is and keep all the request logic in this controller - we can review a bit later. I would encourage you to try to leverage the built in Sortable 'store' approach as it's probably better than building our own. https://github.com/SortableJS/Sortable#store
6. I would keep a reference to the Sortable instance on the controller instance. something like `this.sortable = Sortable(...`.
7. For the `getOptions`, it's better to use `get options() {`, this is well supported in Typescript and is a more intuitive interface in JS classes.
8. We may not need to disable the text selection, I think Sortable will manage that for us (double check).
9. Try to use the table/components framework when adding data attributes. We should not have to manually add any attributes or classses in `wagtail/admin/templates/wagtailadmin/tables/table.html`, instead these should be prepared on the Python side.

**Comment by salty-ivy** [comment]
Ohh! it didn't work as is 🤔 it should be working, I will check it out. Thanks for the feedback will start working on them ASAP.

**Comment by salty-ivy** [comment]
@lb-  I have fixed the dragging,  issues was with `disableTextSelection`.

File: wagtail/admin/templates/wagtailadmin/tables/table.html [hunk 1]
```diff
@@ -13,9 +13,9 @@
             {% endfor %}
         </tr>
     </thead>
-    <tbody>
+    <tbody{% include "wagtailadmin/shared/attrs.html" with attrs=sortable_attr %}>
         {% for row in table.rows %}
-            <tr{% include "wagtailadmin/shared/attrs.html" with attrs=row.attrs %}>
+            <tr{% include "wagtailadmin/shared/attrs.html" with attrs=row.attrs %} >
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/tables/table.html:18]: We should not need changes to this file at all, see main feedback.

File: wagtail/admin/ui/tables/__init__.py [hunk 2]
```diff
@@ -421,6 +424,7 @@ def get_row_classname(self, instance):
 
     def get_row_attrs(self, instance):
         attrs = {}
+        attrs["data-w-sortable-target"] = self.sortable_item_attr
```
Comment (lb-) [inline wagtail/admin/ui/tables/__init__.py:427]: We should not need changes to this file at all if we make the `get_row_attrs` change inside `wagtail/admin/ui/tables/pages.py`. There is already logic to add the row attributes based on the `self.use_row_ordering_attributes`.

**Review by lb- — CHANGES_REQUESTED** [review]
@salty-ivy  - great to meet you this evening, thanks for making the time for the call.

This is now functioning for me, thank you!  As per our chat, here are some additional feedback items. I hope this helps you get to the next stage with this.

## 1 - Simplifying how we set the attributes on the elements & targets

A. Let's get rid of adding the data-controller/value in `wagtail/admin/templates/wagtailadmin/pages/index_results.html` and instead put it into the Python table_kwargs. That code already has logic for attributes for the sorting UI only.

```py
    def get_table_kwargs(self):
        kwargs = super().get_table_kwargs()
        kwargs["use_row_ordering_attributes"] = self.show_ordering_column
        kwargs["parent_page"] = self.parent_page
        kwargs["show_locale_labels"] = self.i18n_enabled and self.parent_page.is_root()
        kwargs["actions_next_url"] = self.get_index_url()

        if self.show_ordering_column:
            kwargs["attrs"] = {
                "aria-description": gettext(
                    "Press enter to select an item, use up and down arrows to move the item, press enter to complete the move or escape to cancel the current move."
                ),
                "data-controller": "w-sortable",
                "data-w-sortable-url-value": "wagtailadmin_pages:set_page_position", # FIX ME - make me into a proper URL with 99999
            }
        return kwargs
```

B. In your controller, we will no longer need the target of `tbody` and can instead use the `this.element` which will be the table. While not as close as tbody, it's good enough for our goals of having the controlled element as close to the elements it controls.

C. We also should not need any changes at all now to `wagtail/admin/templates/wagtailadmin/tables/table.html` as the only attrs we need are on the row.

D. To add the row attribute, instead of passing it down through `wagtail/admin/ui/tables/__init__.py` via access to `self.sortable_item_attr`. Use the existing approach for the other attrs in `wagtail/admin/ui/tables/pages.py` (see the method `get_row_attrs`.

```
    def get_row_attrs(self, instance):
        attrs = super().get_row_attrs(instance)
        if self.use_row_ordering_attributes:
            attrs["id"] = "page_%d" % instance.id
            attrs["data-page-title"] = instance.get_admin_display_title()
            attrs["data-w-sortable-target"] = "item-TEST" # changeme
        return attrs
```

## 2. Refinements to the Stimulus controller

* We should not need `disableTextSelection` at all, this is because changing the `handle` element to a `button` and not a `div` means that the browser will think it's a ... button :) - better for accessibility and avoids workarounds by using the right HTML element. Additionally, I am pretty sure sortable.js will manage this for us.
* We should not need the `tbody` target, as noted above, we can move the controller to the actual table element and access that via `this.element`.
* We will probably still need to revisit the way the fetching is working but for now it's good as is.
* Avoid variables that are specific to the concept of `page`, this code may be used in the future for other reordering. E.g. `pageId` as the variable name can be changed to just `id` or `itemId`.
* Rename `postOrder` to `update`, just a nicer method name. Also change the params to be an object syntax instead of three different params. This makes it clearer when calling it what we are doing and will better support future changes. Something like `{target: HTMLElement, id: string, position: number}`. 
* Rename `getOptions` to `get options` and use like this `this.sortable = Sortable.create(this.element, this.options);`.
* Finally, it would be good to write a Jest unit test for this controller now, the controller code is getting pretty close to final state.

## 3. Accessibility

Great find in the article you shared - https://robbymacdonell.medium.com/refactoring-a-sortable-list-for-keyboard-accessibility-2176b34a07f4 

I think this approach is pretty sound but let's head down the following path and see if this works well. The keyboard interaction could be as follows.

* User tabs through to the first 'handle', when focused on the handle we visibly show two additional bits of info. a. the index number and b. an icon or symbol reflecting that it can be moved up/down such as unicode updown arrow. These should only be visible on focus-visible.
* When focused on one of the handles, pressing up will move the item up, down will move it down. If it's at the ends, the up/down may do nothing and that's OK.
* If the user presses up/down multiple times, we need to consider that and possibly just include a 500ms delay before making the API request but still need to visually move it. We can manage that easily in the Stimulus controller. See how we do it in `client/src/controllers/SwapController.ts` with `searchLazy`. *see below*. This may not work exactly as we want due to us needing to know how many times it's pressed.
* You will need to work out how to use sortable js to move things programmatically, also whatever numbering we show the user will need to be updated to be correct after the move.
* Finally, the focus should be kept on the element that's being moved. This will be the tricky part as when you move a DOM element it's being destroyed and a new one created. So that will need to be accounted for in the code.
* It's important to note that the message that appears at the top of the page (when an update has been made) should still work the same as drag/drop.

```js
# example debounced methods
  connect() {
    this.moveValueChanged = debounce(this.moveValueChanged.bind(this), 500);
    this.sortable = Sortable.create(this.element, this.getOptions());
  }

# ... other methods

  moveValueChanged(value, previousValue) {
    if (this.activeIndex === null || (value === previousValue) { return; }
    this.move();
  }

  move() {
    // this method finalises the manual keyboard move
    // can be enhanced later (one day) to accept DOM events to move items that way
    // however, we will call it on blur (to avoid issues with the debounced keyboard actions) and also call it when the 'moveValueChanged' (debounced) last fires
    // reset the movedValue to zero
    // reset the this.activeIndex to null
    // Do whatever call is needed to 'finalise' the API call here (e.g. call this.postOrder)
  }

  up(event) {
    this.activeIndex = event.target['data-id'??]; // set that we are actively moving an element. This does not need to be a Stimulus value.
    this.movedValue += 1; // create a new value
    // this.sortable.... (find a way to move it visually but not trigger the API call yet)
    // maybe the onEnd: (evt) => { can check for this.moveValue being non-zero and not make the API call.
  }

  down() {
    this.activeIndex = event.target['data-id'??]; // set that we are actively moving an element. This does not need to be a Stimulus value.
    this.movedValue -= 1;
    // this.sortable. .. move down but ensure that the API call does not run.
  }
```

```scss
// client/scss/components/_listing.scss

  .handle {
    cursor: move;
    width: 20px;
    color: theme('colors.icon-secondary');

    &:hover {
      color: theme('colors.icon-secondary-hover');
    }

    &:focus-visible {
      &::before {
        // example of adding a data attribute to be visible only when keyboard focus is active, we could also just use a span.
        content: attr(data-index);
      }
    }
  }
```

**Comment by salty-ivy** [comment]
alright back to the PR, sorry for the dealy!

File: wagtail/admin/templates/wagtailadmin/pages/index.html [hunk 3]
```diff
@@ -30,7 +30,7 @@
         <script defer src="{% versioned_static 'wagtailadmin/js/bulk-actions.js' %}"></script>
     {% endif %}
     <script type="text/javascript">
-        {% if ordering == 'ord' %}
+        {% comment %} {% if ordering == 'ord' %}
```
[the remark this answers is not in the document]
Comment (salty-ivy) [reply wagtail/admin/templates/wagtailadmin/pages/index.html:37]: yes I kept is for reference 😄

**Comment by salty-ivy** [comment]
@lb-  not really sure as its not properly mentioned in the sortable docs but I guess sortable only works with parent child element relation only .

so instead of doing `Sortable.create(this.element, this.options)` I did `Sortable.create(this.element.queySelector('tbody'), this.options)`

as `tr` are the draggable elements and `tbody` is the parent and the former one doesn't seems to work as I noticed every illustration  creates `sortable object` on the direct parent of the draggable items.

File: client/src/controllers/SortableController.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            detail: { clear: true, text, type: 'success' },
+          });
+          document.dispatchEvent(event);
+        })
+        .catch((error) => {
+          console.error('Error:', error);
+        });
+    }
+  }
+
```
Comment (lb-) [inline client/src/controllers/SortableController.ts:60]:
```suggestion

  get containerElement() {
    return this.hasContainerTarget ? this.containerTarget : this.element;
  }

```

File: client/src/controllers/SortableController.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  disableTextSelection(): void {
+    this.element.addEventListener('selectstart', (e) => {
+      e.preventDefault();
+    });
+  }
+
+  connect(): void {
+    this.sortable = Sortable.create(
+      this.element.querySelector('tbody'),
```
Comment (lb-) [inline client/src/controllers/SortableController.ts:85]:
```suggestion
      this.containerElement,
```

File: client/src/controllers/SortableController.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
+import { Controller } from '@hotwired/stimulus';
+import Sortable from 'sortablejs';
+
+export class SortableController extends Controller<HTMLElement> {
+  static targets = ['handle', 'item'];
```
Comment (lb-) [inline client/src/controllers/SortableController.ts:5]:
```suggestion
  static targets = ['container', 'handle', 'item'];
```

**Review by lb- — COMMENTED** [review]
Great investigation about how Sortable works. I'd recommend we do add a target for the sorting container (like you originally did). However, let's make it a generic name like 'container', instead of specific 'tbody'. 

This also avoids you hard-coding a selector. While this may not be used outside of this page for now it's nice to try not to hard code assumed Dom structure if we can. 

Here's a recommendation for how to make this change.

**Comment by lb-** [comment]
I've got two other ideas for the parent element approach. 

1. Just get the parent of the first item target. 
```js
const parentElement = this.itemTarget.parentElement;
```

This is way simpler and also will throw an error (which is probably good) when there are no item targets declared. 


2. Accept a selector value. 

Add a value of `container' as a string with a default of an empty string. 

Then you get the value like this

```js
const containerElement = this.containerValue ? this.element.querySelector(this.containerValue) : this.element;
```

This second approach is still quite simple and gives a little more flexibility to how the controller can be used.

**Comment by salty-ivy** [comment]
@lb-  how about combining the 2 approaches to single conditional return so that this can be used in other  places and won't be specific to this controller only
```ts
// defining a container values

container: { default: "", type: String },

declare containerValue: string;

// a getter type function

getContainerElement(container: string | null = null): HTMLElement | null{
    if (container){
      this.containerValue = container;
      return this.containerValue ? this.element.querySelector(this.containerValue) : this.element;
    }else{
      return this.itemTarget.parentElement;
    }
  }
```

**Comment by lb-** [comment]
@salty-ivy I think that may be overcomplicating it a bit, our primary goal is to avoid hard-coding `'tbody'` anywhere in our controller.

As it may be tricky to add an attribute directly to the `tbody` element I think passing in a selector is probably simplest.

Additionally, we want to avoid writing to a Stimulus value of a string with anything other than a string, this will not work as you expect. The 'setter' on this will write to the data attribute and you will probably end up with an attribute value of `'[Object object]'`.

So, how about we just have a `container` value, it's expected to be a selector string. If it's empty we just assume that the controlled element is the container. No need for a new method, we can just do this in one line.

```
  const containerValue = this.containerValue;
 this.sortable = Sortable.create(
      ((containerValue && document.querySelector(containerValue)) || this.element) as HTMLElement,
      this.options,
    );
```

Something like that could do the trick, this means you only need one additional attribute where you set the data-controller, `"data-w-sortable-container-value": "tbody"`.

If we need to solve for more complex cases (like nested sortable controllers) we can address that in a future bit of work.

**Comment by salty-ivy** [comment]
I see that's great!

**Comment by lb-** [comment]
@salty-ivy that makes sense, give that a go, push the code up and let's take a look. Thanks again for pursuing this.

**Comment by salty-ivy** [comment]
@lb-  I have put some draft functional code that works with keyboard let me know your thoughts.

**Comment by salty-ivy** [comment]
also changing the handle from div to `button` changes its visual look.

**Comment by salty-ivy** [comment]
https://github.com/wagtail/wagtail/assets/74553951/f75d0a00-e3f2-482c-9603-dd5658e1fef4

**Comment by salty-ivy** [comment]
https://github.com/wagtail/wagtail/assets/74553951/e18dd56f-d189-400a-9709-e48c628d1bb2

**Comment by lb-** [comment]
Awesome. I'll take a look.

**Comment by salty-ivy** [comment]
https://github.com/wagtail/wagtail/assets/74553951/43e8143b-9f29-4645-bae0-cff0ee0b2c44

**Comment by salty-ivy** [comment]
@lb-  Instead making use of debounce and call at blur of the `move` function why don't inspire it a bit from the previous keyboard controls, where to enter and exit we needed to press `enter` same with the new controls user can play around with the keyboard control and finally pressing `enter` to make the changes permanent.

as show on the above recording.

**Comment by salty-ivy** [comment]
also I am not sure whether `button` and `sortable` is helping with text disabling at all?

**Comment by lb-** [comment]
Sorry - had to fix some classes.

**Review by lb- — APPROVED** [review]
Screen reader validation screenshots for reference.

Screenshot 2024-01-19 at 7 52 18 am

Screenshot 2024-01-19 at 7 52 07 am

**Comment by lb-** [comment]
New issue up to add support for `caption` elements in the tables UI component, this will mean we do not have to use aria-description. 

https://github.com/wagtail/wagtail/issues/11493

**Pull request merged**